Fix typo CXX_FLAGS -> CXXFLAGS in ASAN script and docu

Flamefire opened the pull request:
The correct environment variable is CXXFLAGS not CXX_FLAGS which likely was influenced by CMakes naming: `CMAKE_CXX_FLAGS`. See e.g. https://en.wikipedia.org/wiki/CFLAGS

Flamefire committed:
Fix typo CXX_FLAGS -> CXXFLAGS in ASAN script and docu

mrshenli commented:
Test failures look relevant?

```
Jul 23 17:41:27 + python -c 'import torch; torch._C._crash_if_csrc_asan(3)'
Jul 23 17:41:29 /var/lib/jenkins/workspace/torch/csrc/Module.cpp:146:3: runtime error: index 3 out of bounds for type 'volatile char [3]'
Jul 23 17:41:29     #0 0x7f3457fd65b0 in THPModule_crashIfCsrcASAN(_object*, _object*) (/opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so+0x48ec5b0)
Jul 23 17:41:29     #1 0x564fbd13e2c4 in _PyCFunction_FastCallDict /tmp/build/80754af9/python_1588903631989/work/Objects/methodobject.c:209
Jul 23 17:41:29     #2 0x564fbd1c5b0b in call_function /tmp/build/80754af9/python_1588903631989/work/Python/ceval.c:4851
```

Flamefire commented:
Those are expected failures. As from the name of the functions: They are supposed to trigger ASAN or UBSAN and it seems they do.

The actual failure is different:
- `python -c 'import torch; torch._C._crash_if_csrc_ubsan(0)'` is run
- ` SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /var/lib/jenkins/workspace/torch/csrc/Module.cpp:155:31 in ` is detected
- `retcode=0` is returned which should have been `retcode=1` due to the UBSAN triggered above

BTW: When trying this locally I got lot's of undefined references into the ASAN library when it tries to link some C++ binaries. The documentation (and test) uses only `CFLAGS="-fsanitize=address -fno-sanitize-recover=all -shared-libasan`" but not the same for `CXXFLAGS`. It likely works on CI because CMake does add some `CMAKE_CXX_FLAGS_DEBUG` when `USE_ASAN=1`, however this makes it fail locally to build Google.Benchmark, see https://github.com/pytorch/pytorch/issues/41916

Maybe the Docker environment already has some CXXFLAGS set?

Questions:
- May I add a commit to this PR which adds to the documentation to set both `CFLAGS` and `CXXFLAGS` instead of only `CFLAGS` (I tested it and setting CXXFLAGS makes it work)
- Where is `build-asan.sh` actually called in those failing pipelines? I don't see any build process so it's hard to tell what is or is not set.
- Similar to the first point I could add the asan flags to CXXFLAGS to in build-asan.sh for CI and expect it to pass. Likely the `-fno-sanitize-recover=all` is missing or gets lost. Shall I do that?
- To debug, I'd suggest to print out the contents of CFLAGS/CXXFLAGS on invocation of build-asan.sh on CI. Maybe we need to include the original flags. But to tell for sure I'd need to know/see where the script is invoked. Any ideas?

Flamefire committed:
Also add sanitizer flags to CXXFLAGS

This avoids undefined references during linking of C++ binaries that
contain C object files and also failures on CI as detected UBSAN will
not lead to an non-null exit code as -fno-sanitize-recover=all is
missing for C++ files

Flamefire commented:
Ok, CXXFLAGS didn't contain any value: https://circleci.com/api/v1.1/project/github/pytorch/pytorch/6352346/output/106/0?file=true&allocation-id=5f1a81a73eb53b5129da43ad-0-build%2F24A92511

So I don't know why this PR would introduce a new failure but I'd rather expect that it would have failed already.

Trying the "proper" fix with setting the CXXFLAGS as mentioned above now.

Flamefire commented:
And I can report success: The new build shows (as expected):
```
+ python -c 'import torch; torch._C._crash_if_csrc_ubsan(0)'
 SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /var/lib/jenkins/workspace/torch/csrc/Module.cpp:154:18 in 
+ retcode=1
```

Compare https://app.circleci.com/pipelines/github/pytorch/pytorch/192802/workflows/02ea860b-4156-46ae-a669-cb3405b6de63/jobs/6339477 (prior to the 2nd commit) and https://app.circleci.com/pipelines/github/pytorch/pytorch/193146/workflows/b261886a-7885-43c2-b4e6-c90765be8ba2/jobs/6353530 (after that) which shows that the change works as expected.

malfet commented on .jenkins/pytorch/build-asan.sh:
```diff
@@ -32,7 +32,7 @@ export ASAN_OPTIONS=detect_leaks=0:symbolize=1:detect_odr_violation=0
 CC="clang" CXX="clang++" LDSHARED="clang --shared" \
   CFLAGS="-fsanitize=address -fsanitize=undefined -fno-sanitize-recover=all -shared-libasan -pthread" \
-  CXX_FLAGS="-pthread" \
+  CXXFLAGS="-fsanitize=address -fsanitize=undefined -fno-sanitize-recover=all -shared-libasan -pthread" \
```
Specifying any of the options shouldn't be necessary, as `USE_ASAN` CMakeList option should to the trick

malfet commented on CONTRIBUTING.md:
```diff
@@ -857,7 +857,7 @@ build_with_asan()
   LDFLAGS="-stdlib=libstdc++" \
   CFLAGS="-fsanitize=address -fno-sanitize-recover=all -shared-libasan -pthread" \
-  CXX_FLAGS="-pthread" \
+  CXXFLAGS="-fsanitize=address -fno-sanitize-recover=all -shared-libasan -pthread" \
```
This shouldn't be necessary, but instead we should pass `USE_ASAN=1` environment variable.

malfet reviewed (changes requested):
Fixing typo (`CXX_FLAGS`->`CXXFLAGS`) is indeed the right thing, but we should not be passing any sanitizer flags explicitly but rather define all ASAN specific flags in one location, namely here:
https://github.com/pytorch/pytorch/blob/dede71d6e31e259698ce4745070db99149402ec3/CMakeLists.txt#L595-L596

Flamefire commented:
While I agree that using CMake instead makes things easier I'd rather like to do this in 2 steps:
1. Fix the sanitizer build in general which is suggested to (and run like that on CI) via environment variables. As reported in https://github.com/pytorch/pytorch/pull/41925#issuecomment-663368905 the same flags passed to CFLAGS must be passed to CXXFLAGS or builds will fail due to different ASAN libs used and missing the "fail" flag
2. Once this is working one can tackle the "TODO" mentioned at https://github.com/pytorch/pytorch/pull/41925/files#diff-e6488795ad4e0b25afd2720b38d63addR32

However step 2 is a lot more work than it might seem. See my comments in https://github.com/pytorch/pytorch/issues/41916 for an analysis of open things to fix.

> Specifying any of the options shouldn't be necessary, as USE_ASAN CMakeList option should to the trick

The CMake option enables significantly less.
- Env var: `-fsanitize=address -fsanitize=undefined -fno-sanitize-recover=all -shared-libasan -pthread`
- CMake var: `-fsanitize=address`

So your are missing UBSAN, fail-on-error, the shared ASAN lib and the `-pthread` workaround in CMake. So for now until #41916 is addressed, I'd very much prefer if we fix the usage with CFLAGS/CXXFLAGS first as that is way easier

Flamefire commented:
See open comment https://github.com/pytorch/pytorch/pull/41925#issuecomment-663827603

Flamefire commented:
No stale but needs rebasing and that needs defining a strategy to taggle the multiple mentioned issues.
